### PR TITLE
debezium/dbz#1554 Fix data loss on restart when multiple events share the same LSN

### LINKED ISSUE #1554: Postgres Source with jdbc copy will lost data.
## Bug report

**What Debezium connector do you use and what version?**

3.4

---

**What is the connector configuration?**

default config

---

**What is the captured database version and mode of deployment?**

any version

**What behavior do you expect?**
When insert with jdbc copy( using mutl-insert will merger multiple message into one lsn is always true: 

Image

For example:
```sql
psql -h 127.0.0.1 -p5438 -U postgres -d postgres  -c "\copy public.lsn_test(id,val) FROM '//mnt/disk1/wudi/test.csv' WITH (FORMAT csv, HEADER false)
```

Image

We can see the lsn of multiple message is same. While isLastEventForLsn of PgOutputReplicationMessage is always true. If the job is stopped at the first one, then restarted. Data will be lost.

---

**What behavior do you see?**

Postgres Source loses data.

---

**Do you see the same behaviour using the latest released Debezium version?**

_Ideally, also verify with the latest Alpha/Beta/CR version._

`<Your answer>`

---

**Do you have the connector logs, ideally from start till finish?**

_You might be asked later to provide [DEBUG/TRACE](https://debezium.io/documentation/reference/stable/operations/logging.html) level log._

`<Your answer>`

---

**How to reproduce the issue using our [tutorial](https://github.com/debezium/debezium-examples/tree/main/tutorial) deployment?**

psql -h 127.0.0.1 -p5438 -U postgres -d postgres  -c "\copy public.lsn_test(id,val) FROM '//mnt/disk1/wudi/test.csv' WITH (FORMAT csv, HEADER false)

### PR DESCRIPTION
**Pull request by ganeshannt**
Fixes debezium/dbz#1554

## Description

When using PostgreSQL COPY or batched multi-row inserts, multiple DML events can share
the same LSN within a transaction. Previously, PgOutputReplicationMessage.isLastEventForLsn()
always returned true, causing the connector to mark the LSN as fully processed after the
first event. On restart, remaining events at that LSN would be skipped, resulting in data loss.

This fix introduces a per-LSN event counter in the connector offsets that tracks how many
events at the current LSN have been successfully processed. On restart, WalPositionLocator
uses this counter to resume precisely after the last processed event within a same-LSN group.

Changes:
- PgOutputReplicationMessage.isLastEventForLsn() now returns false for DML events
- Added lsn_events_processed field to PostgresOffsetContext (backward compatible, defaults to 0)
- Updated WalPositionLocator to use LSN + event count for resume point calculation
- PostgresStreamingChangeEventSource increments counter per DML, resets on COMMIT
- Added unit tests for same-LSN multi-event restart scenarios

## PR Checklist

- [x] I have read the [contribution guidelines](https://github.com/debezium/debezium/blob/main/CONTRIBUTING.md) and the [governance document](https://github.com/debezium/governance/blob/main/GOVERNANCE.md) on PR expectations.
- [x] Minimal changes to code not directly related to your change (e.g. no unnecessary formatting changes or refactoring to existing code)
- [x] One feature/change per PR unless tightly coupled
- [x] Do a rebase on upstream `main`

### CODE REVIEW

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java [hunk 1]
```diff
@@ -344,6 +344,82 @@ void testToStringAfterProcessingMessages() {
                 .contains("startStreamingLsn=" + lsn);
     }
 
+    // =============================================================================
+    // Same-LSN Event Counter Tests (dbz#1554)
+    // =============================================================================
+
+    @Test
+    void whenMultipleEventsShareSameLsnShouldResumeAfterProcessedCount() {
```
Comment (Naros) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java:350]:
```suggestion
    @Test
    @FixFor("debezium/dbz#1554")
    void whenMultipleEventsShareSameLsnShouldResumeAfterProcessedCount() {
```

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java [hunk 2]
```diff
@@ -344,6 +344,82 @@ void testToStringAfterProcessingMessages() {
...
+        assertThat(result).isEmpty();
+
+        // Next LSN arrives - should resume from here
+        result = locator.resumeFromLsn(nextLsn, insertMessage);
+        assertThat(result).isPresent();
+        assertThat(result.get()).isEqualTo(nextLsn);
+    }
+
+    @Test
+    void whenLsnEventsProcessedIsZeroShouldReplayAllEventsAtThatLsn() {
```
Comment (Naros) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java:424]:
```suggestion
    @Test
    @FixFor("debezium/dbz#1554")
    void whenLsnEventsProcessedIsZeroShouldReplayAllEventsAtThatLsn() {
```

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java [hunk 3]
```diff
@@ -344,6 +344,82 @@ void testToStringAfterProcessingMessages() {
                 .contains("startStreamingLsn=" + lsn);
     }
 
+    // =============================================================================
+    // Same-LSN Event Counter Tests (dbz#1554)
+    // =============================================================================
```
Comment (Naros) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/junit/WALPositionLocatorTest.java]: These comments are superfluous, can you remove them.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java [hunk 4]
```diff
@@ -45,6 +46,8 @@ public class PostgresOffsetContext extends CommonOffsetContext<SourceInfo> {
     private Lsn streamingStoppingLsn = null;
     private final TransactionContext transactionContext;
     private final IncrementalSnapshotContext<TableId> incrementalSnapshotContext;
+    private long lsnEventsProcessed;
```
Comment (vjuranek) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:49]: If I'm not mistaken, you don't set `lsnEventsProcessed` anywhere. How does it survive Debezium restart?

Comment (ganeshannt) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:49]: Good catch for asking! It's actually persisted in getOffset() where result.put(LSN_EVENTS_PROCESSED_KEY, lsnEventsProcessed) writes it to the offset map, and restored in Loader.load() via getOrDefault(LSN_EVENTS_PROCESSED_KEY, 0L). 
default 0 keeps it backward compatible with old offsets.

Comment (vjuranek) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:49]: Thanks @ganeshannt , I completely missed that.

**Comment by vjuranek** [comment]
@ganeshannt please rebase also this PR on the latest main to avoid Apicurio failures. Thanks!

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java (hunk 4 above)
Comment (ganeshannt) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:49]: no worries 👍🏼

**Comment by ganeshannt** [comment]
@vjuranek , Rebased on latest main

**Comment by jpechane** [comment]
@ganeshannt Test failures seems to be related.

**Comment by ganeshannt** [comment]
> @ganeshannt Test failures seems to be related.

@jpechane Updated the approach to handle this properly.

The previous approach broke `lastCompletelyProcessedLsn` tracking by changing `isLastEventForLsn()` to return false for DML events. Also, `skipMessage()` filtered by LSN only, so unprocessed events sharing the same LSN got filtered during replay.

Updated logic reverts `isLastEventForLsn()` back to true, and instead teaches `WalPositionLocator` to set the resume point at the stored LSN with a skip count. `skipMessage()` now counts down processed events before passing, so same-LSN events that weren't processed don't get lost. Tests cover both search and replay phases now. Let me know if you have any other thoughts.

**Comment by ganeshannt** [comment]
@jpechane, Pushed a fix for the CI failures. issue was the skip threshold so changed from `> 0` to `> 1` in `WalPositionLocator.resumeFromLsn()`.
In pgoutput, BEGIN and the first DML can share the same LSN. With `lsnEventsProcessed = 1` (the normal single-event case), the search phase counted both as 2 events at the stored LSN, exceeded the threshold, and incorrectly activated the skip path causing one record to be lost on restart.
Fix only activates same-LSN skip counting when `lsnEventsProcessed > 1`, meaning there were genuinely multiple DML events at that LSN (COPY/batch scenario). For counter 0 or 1, the existing "resume from next different LSN" logic handles restart correctly. Added a test for the single-event case.

correct me if i missing any scenario here

**Comment by ganeshannt** [comment]
@jpechane @Naros @vjuranek  - CI is green now. Ready for review when you get a chance

**Review by vjuranek — APPROVED** [review]
LGTM, but as it's a change with potentially big impact, let's wait for more reviews.

**Review by Naros — APPROVED** [review]
LGTM, just one question @ganeshannt if its possible to have an integration test in addition to the WALPositionLocator logic in isolation, or is that a bit difficult to do?

**Comment by ganeshannt** [comment]
> LGTM, just one question @ganeshannt if its possible to have an integration test in addition to the WALPositionLocator logic in isolation, or is that a bit difficult to do?

@Naros, Yeah, it’s good to add integration tests too. I added edge case unit tests and an integration test as requested. Unit tests cover the boundary at lsnEventsProcessed=2, all events already processed at stored LSN, and correct skip counter decrement. Integration test inserts 20 rows in a single transaction, stops mid-batch, restarts, and verifies no data loss. Pgoutput-only.

**Comment by ganeshannt** [comment]
@Naros @vjuranek - I’ve added the integration test fixes. can you trigger the CI pipeline

**Comment by ganeshannt** [comment]
@vjuranek @Naros ,  All PostgreSQL, JDBC, Apicurio, and Core Component jobs pass. The only failing job is "Debezium Platform - conductor", which fails resolving debezium-schema-generator:3.6.0-SNAPSHOT while building debezium-connector-db2. This is unrelated to the changes in this PR. 
Could a maintainer retry that job?

**Comment by vjuranek** [comment]
@ganeshannt Debezium Platform failure is fine, it's obviously unrelated. IIRC @Naros is fine with the change, so IMHO this is ready to be merge. However, as we are going to release 3.6.0.Final tomorrow and as I wrote, this change has potentially big impact, I'd suggest merge it right after 3.6.0.Final (later tomorrow or on Wed) and release in 3.7.0.Alpha1.

**Comment by vjuranek** [comment]
Merged, thanks @ganeshannt !
